### DISCUSSION TITLE
time.sleep() accuracy

### DISCUSSION THREAD

**Discussion by Dandy-Don** [micropython/micropython#14444] [RP2040 / Pico]
Hi, everyone!

I am having trouble getting good delays out of my Pico using Micropython v1.20....

The code is simple.  Just checking a pin to see when it gets pulled to ground.  Then set another pin low, delay for a few milliseconds, then set that pin high again.

```
from machine import Pin
import time

led = Pin(25, Pin.OUT)
XOR = Pin(14, Pin.OUT)
switch = Pin(15, Pin.IN, Pin.PULL_UP)

XOR.on()

while XOR:
    led.toggle()
    time.sleep(1)

XOR.off()
time.sleep(0.01)
XOR.on()
```

When I run this I get crazy results.  The last `time.sleep(0.01)` delay never gives me the same delay twice.  I get mixed results...

I tried using an interrupt handler to see if that helps, but nothing changed.

```
def sw_isr(pin):
    XOR.off()
    time.sleep(0.01)
    XOR.on()

switch.irq(trigger=Pin.IRQ_FALLING, handler=sw_isr)
```

Suggestions?  Any help will be greatly appreciated!
--Don

**Comment by robert-hh** [comment]
I wonder if the loop below ever terminates, since XOR is a Pin object, which always evaluates to True.
```
while XOR:
    led.toggle()
    time.sleep(1)
```
Instead of using time.sleep(0.01) you could try time.sleep_ms(10).

**Reply by Dandy-Don** [reply]
I don't know if the loop terminates or not.  I know that it is jumping out of the loop to the next steps then coming back into to the loop...  I see that on an oscilloscope.

Or does the processor 'reset' at the end of the code?

**Reply by robert-hh** [reply]
> Or does the processor 'reset' at the end of the code?

No. It would usually just fall back to REPL. But if you use Thonny, it might. But something is weird with your code: From the description, I had expected that the loop should be:
```
while switch():
    led.toggle()
    time.sleep(1)
```
`switch()` returns the level at Pin 15
Edit: You can as well use XOR(0) and XOR(1) instead of XOR.off() and XOR.on(). That's a little bit faster.

**Reply by Dandy-Don** [reply]
No.  The loop is waiting for XOR to go low, then it jumps out of the loop....

**Reply by robert-hh** [reply]
Please look at your code posted above. XOR is created in Pin.OUT mode, with not value set. In that mode the Pin will be driven actively high or low, depending on the value being set. Here the Pin will be driven low. In the loop, you write `while XOR:`. This will NOT test the level at the XOR pin. If you want to read the level, you have to write `XOR()` or `XOR.value()` (without argument). But since XOR is set to output mode, that will only return the level set by a previous write command.

**Reply by Dandy-Don** [reply]
Yep.  I figured it out.....  I don't know what I was thinking.  It seems to be working better now.  Thanks!

**Comment by bixb922** [comment]
`time.sleep_us(10_000)` does a busy wait but timing is more precise than sleep_ms

**Reply by Dandy-Don** [reply]
Thanks!  I will try that.
